### ISSUE TITLE
API: support batch write

### ISSUE THREAD

**Issue by drmingdrmer** [datafuselabs/openraft#169]
1. `openraft::raft::RaftMsg::ClientWriteRequest`:
Change `rpc` to a `Vec` to contain more than one client request:

```rust
    ClientWriteRequest {
        rpc: ClientWriteRequest<D>,
        tx: RaftRespTx<ClientWriteResponse<R>, ClientWriteError>,
    },
```

2. Update API `client_write()` to support submitting more than one log at a time.

Considerations:
- Is it legal to have zero-length `Vec` in a `ClientWriteRequest`? If it is legal, maybe it has to be dealt with: because it does actually append any log, and a reply should be sendt to `tx` at once.
- `tx` is sent only when the last log in `rpc` is applied. Or there should be a `tx` for every `ClientWriteRequest<D>`?

**Comment by schreter** [comment]
Just a note: please no `Vec<T>`. We need to get rid of them in all places in favor of something zero-cost, like `impl IntoIterator<T>` or similar, since a `Vec` is just an unnecessary memory allocation.

I personally dislike `Vec`, `String` and similar, since these allocations are variable-sized. This is in general causing fragmentation in the memory manager, which leads to slowdowns and sooner or later to out-of-memory errors due to excessive fragmentation. Having written a memory manager for an in-memory database (in C++), I had my share of problems in this respect :-).

**Comment by drmingdrmer** [comment]
@schreter 
The `rpc` is a field of the following `enum`. Does It have to have another generic type argument if we use the trait `IntoIterator<T>` to describe the type of `rpc`?
```rust
pub(crate) enum RaftMsg<D: AppData, R: AppDataResponse> {
    ...
    ClientWriteRequest {
        rpc: ClientWriteRequest<D>,
        tx: RaftRespTx<ClientWriteResponse<R>, ClientWriteError>,
    },
```


> Just a note: please no `Vec<T>`. We need to get rid of them in all places in favor of something zero-cost, like `impl IntoIterator<T>` or similar, since a `Vec` is just an unnecessary memory allocation.

If it is possible to use some generic type, why not just use `impl Iterator<T>`?




> I personally dislike `Vec`, `String` and similar, since these allocations are variable-sized. This is in general causing fragmentation in the memory manager, which leads to slowdowns and sooner or later to out-of-memory errors due to excessive fragmentation. Having written a memory manager for an in-memory database (in C++), I had my share of problems in this respect :-).

:DD

**Comment by schreter** [comment]
> Does It have to have another generic type argument if we use the trait IntoIterator<T> to describe the type of rpc?

You are right, it is not optimal to do it now. After the refactoring of the generics to a single config type it will be much easier.
